module/zfs/dbuf.c

```diff
@@ -2701,28 +2753,154 @@ dmu_buf_is_dirty(dmu_buf_t *db_fake, dmu_tx_t *tx)
+	 * unauthenticated blocks, which will verify their MAC if
+	 * the key is now available.
+	 */
+	if (err == 0 && db->db_buf != NULL &&
+	    (arc_is_encrypted(db->db_buf) ||
+	    arc_is_unauthenticated(db->db_buf) ||
+	    arc_get_compression(db->db_buf) != ZIO_COMPRESS_OFF)) {
+		zbookmark_phys_t zb;
+
+		SET_BOOKMARK(&zb, dmu_objset_id(db->db_objset),
+		    db->db.db_object, db->db_level, db->db_blkid);
+		dbuf_fix_old_data(db, spa_syncing_txg(spa));
+		err = arc_untransform(db->db_buf, spa, &zb, B_FALSE);
+		dbuf_set_data(db, db->db_buf);
+	}
+	DB_DNODE_EXIT(db);
+	DBUF_STAT_BUMP(hash_hits);
+
+	return (err);
+}
+
 void
-dmu_buf_will_clone(dmu_buf_t *db_fake, dmu_tx_t *tx)
+dmu_buf_will_clone_or_dio(dmu_buf_t *db_fake, dmu_tx_t *tx)
 {
 	dmu_buf_impl_t *db = (dmu_buf_impl_t *)db_fake;
 
 	/*
-	 * Block cloning: We are going to clone into this block, so undirty
-	 * modifications done to this block so far in this txg. This includes
-	 * writes and clones into this block.
+	 * Block clones and Direct I/O writes always happen in open-context.
 	 */
+	ASSERT(!dmu_tx_is_syncing(tx));
+	ASSERT(db->db_level == 0);
+	ASSERT(db->db_blkid != DMU_BONUS_BLKID);
+	ASSERT(db->db.db_object != DMU_META_DNODE_OBJECT);
+
 	mutex_enter(&db->db_mtx);
 	DBUF_VERIFY(db);
-	VERIFY(!dbuf_undirty(db, tx));
+
+	/*
+	 * We are going to clone or issue a Direct I/O write on this block, so
+	 * undirty modifications done to this block so far in this txg. This
+	 * includes writes and clones into this block.
+	 *
+	 * If there dirty record associated with this txg from a previous Direct
+	 * I/O write then space accounting cleanup takes place. It is important
+	 * to go ahead free up the space accounting through dbuf_undirty() ->
+	 * dbuf_unoverride() -> zio_free(). Space accountiung for determining
+	 * if a write can occur in zfs_write() happens through dmu_tx_assign().
+	 * This can cuase an issue with Direct I/O writes in the case of
+	 * overwriting the same block, because all DVA allocations are being
+	 * done in open-context. Constantly allowing Direct I/O overwrites to
+	 * the same block can exhaust the pools available space leading to
+	 * ENOSPC errors at the DVA allocation part of the ZIO pipeline, which
+	 * will eventually suspend the pool. By cleaning up sapce acccounting
+	 * now, the ENOSPC error can be avoided.
+	 *
+	 * Since we are undirtying the record in open-context, we must have a
+	 * hold on the db, so it should never be evicted after calling
+	 * dbuf_undirty().
+	 */
+	VERIFY3B(dbuf_undirty(db, tx), ==, B_FALSE);
 	ASSERT0P(dbuf_find_dirty_eq(db, tx->tx_txg));
+
 	if (db->db_buf != NULL) {
-		arc_buf_destroy(db->db_buf, db);
+		/*
+		 * If there is an associated ARC buffer with this dbuf we can
+		 * only destroy it if  the previous dirty record does not
+		 * reference it.
+		 */
+		dbuf_dirty_record_t *dr = list_head(&db->db_dirty_records);
+		if (dr == NULL || dr->dt.dl.dr_data != db->db_buf)
+			arc_buf_destroy(db->db_buf, db);
```

Comment:
I think we could upstream this chunk ASAP separately from the rest of the PR, since it seems like a block cloning issue, that @ixhamza may have just hit in his other work.

Reply:
Fine by me. I will create a PR that updates `dmu_buf_will_clone()` with this logic added. @ixhamza any chance you have a test case we could put together to add to the ZTS test suite?

Reply:
PR has been opened #16337. Happy to add a ZTS test case as well to the PR if we have an example to work with.

Reply:
@bwatkinson - I don't have a reproduction scenario with the file cloning at the moment, but I will test more.

Reply:
@bwatkinson - I was able to reproduce the panic with the following script (running for a couple of minutes) without [your patch](https://github.com/openzfs/zfs/pull/16337) applied. After applying [your patch](https://github.com/openzfs/zfs/pull/16337),  I can not reproduce the issue after running the script for an hour:
```
#!/bin/sh
zpool destroy tank
zpool create tank nvme0n1 -O mountpoint=/tank
zfs create -o sync=always -o recordsize=4K tank/d1
zfs create -o sync=always -o recordsize=4K tank/d2
dd if=/dev/urandom of=/tank/d1/file bs=1M count=1k status=progress
zpool sync
while true
do
	/usr/share/zfs/zfs-tests/bin/clonefile -f /tank/d1/file /tank/d2/file 0 0 1073741824
	zpool sync
done &
while true
do
	dd if=/dev/urandom of=/tank/d2/file bs=1M count=1k
done &
```

Reply:
Alright, I just updated the PR to include this test case in the ZTS `block_cloning` tests.